### PR TITLE
adds properties to dismissible widget

### PR DESCRIPTION
**Pull request by theankurkedia**
Adds 2 properties.
`crossAxisEndOffset` which defines the end offset across the main axis after the card is dismissed.
`rollbackDuration` which defines the duration for card to come back to original position if not dismissed.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 1]
```diff
@@ -1,7 +1,8 @@
-// Copyright 2015 The Chromium Authors. All rights reserved.
-// Use of this source code is governed by a BSD-style license that can be
-// found in the LICENSE file.
+// Copyright 2015 The Chromium Authors. All rights reserved.		
+// Use of this source code is governed by a BSD-style license that can be		
+// found in the LICENSE file.		
 
+ 
 import 'package:flutter/foundation.dart';
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/dismissible.dart]: looks like you have some accidental whitespace changes near the top (new spaces and new lines)

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 2]
```diff
@@ -134,6 +136,13 @@ class Dismissible extends StatefulWidget {
   /// [direction] property.
   final Map<DismissDirection, double> dismissThresholds;
 
+  /// Defines the duration for card to come back to original position if not dismissed.
+  final Duration rollbackDuration;
+
+  /// Defines the end offset across the main axis after the card is dismissed.
+  /// If non-zero value is given then widget moves in cross direction depending on whether it is positive or negative.
+  final double crossAxisEndOffset;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/dismissible.dart]: Add a blank line in the documentation after the first sentence (the first paragraph is used when we need a short description).

**Comment by Hixie** [comment]
This looks good. One thing it will need is tests, so that we don't accidentally break this or remove the feature at some point. Anything you care about making sure we don't break, please add a test to verify that it works. You can add the tests to: https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/dismissible_test.dart

**Comment by theankurkedia** [comment]
@Hixie I have added the tests. Both the properties are being covered in the code.

File: packages/flutter/test/widgets/dismissible_test.dart [hunk 3]
```diff
@@ -143,6 +147,58 @@ Future<Null> dismissItem(WidgetTester tester, int item, {
...
+      // edge and outside the Dismissible event listener's bounds.
+      downLocation = tester.getBottomLeft(finder) + const Offset(0.0, -0.1);
+      upLocation = tester.getBottomLeft(finder) + const Offset(0.0, -0.2);
+      break;
+    case AxisDirection.down:
+      // again with doing the same here for symmetry
+      downLocation = tester.getTopLeft(finder) + const Offset(0.0, 0.1);
+      upLocation = tester.getTopLeft(finder) + const Offset(0.0, 0.2);
+      break;
+    default:
```
Comment (Hixie) [inline packages/flutter/test/widgets/dismissible_test.dart]: you can remove the default; the analyzer will then warn us at analyze time if there's an unsupported gesture direction

File: packages/flutter/test/widgets/dismissible_test.dart [hunk 4]
```diff
@@ -143,6 +147,58 @@ Future<Null> dismissItem(WidgetTester tester, int item, {
...
 }
 
+Future<Null> rollbackElement(WidgetTester tester, Finder finder, { @required AxisDirection gestureDirection }) async {
+  Offset downLocation;
+  Offset upLocation;
+  switch (gestureDirection) {
+    case AxisDirection.left:
+      // getTopRight() returns a point that's just beyond itemWidget's right
+      // edge and outside the Dismissible event listener's bounds. Just nudge
+      // the itemWidget before it's threshold.
```
Comment (Hixie) [inline packages/flutter/test/widgets/dismissible_test.dart]: its

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 5]
```diff
@@ -83,6 +82,8 @@ class Dismissible extends StatefulWidget {
     this.direction: DismissDirection.horizontal,
     this.resizeDuration: const Duration(milliseconds: 300),
     this.dismissThresholds: const <DismissDirection, double>{},
+    this.rollbackDuration: const Duration(milliseconds: 300),
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/dismissible.dart]: this should be 200ms, right? To match the previous _kDismissDuration?

File: packages/flutter/test/widgets/dismissible_test.dart [hunk 6]
```diff
@@ -558,4 +614,29 @@ void main() {
...
+
+    await rollbackItem(tester, 0, gestureDirection: AxisDirection.down);
+    expect(find.text('0'), findsOneWidget);
+    expect(dismissedItems, isEmpty);
+
+    await rollbackItem(tester, 1, gestureDirection: AxisDirection.up);
+    expect(find.text('1'), findsOneWidget);
+    expect(dismissedItems, isEmpty);
+  });
+}
```
Comment (Hixie) [inline packages/flutter/test/widgets/dismissible_test.dart]: maybe i misunderstand the test but it doesn't seem to me that it verifies that things actually go diagonally or that they actually take rollbackDuration time to go away. For example, if someone replaced `widget.crossAxisEndOffset` with `0.0` in `_updateMoveAnimation`, would the test fail? Similarly, would your test fail if the duration was not precisely honoured, but instead e.g. the rollback happened in half the specified time? or double the specified time?

File: packages/flutter/lib/src/widgets/dismissible.dart (hunk 5 above)
Comment (theankurkedia) [reply packages/flutter/lib/src/widgets/dismissible.dart]: Yes, you are right.

File: packages/flutter/test/widgets/dismissible_test.dart (hunk 6 above)
Comment (theankurkedia) [reply packages/flutter/test/widgets/dismissible_test.dart]: You are right again, what all it does is check if the widget is still there after giving it a small nudge. For `rollbackDuration`, I am trying to find a more specific way to test the features but I am caught up in it. What I have in mind now is to check the duration once the widget is released from drag, but I can't seem to find a way to do it. And for `crossAxisEndOffset`, I am trying to get the `topLeft` or `topRight` offset and check, but I am not sure if it is the right way to go ahead. I haven't had success doing it yet.

Comment (Hixie) [reply packages/flutter/test/widgets/dismissible_test.dart]: For the duration, you can just pump a duration less than the rollback, verify it hasn't finished rolling back, then pump a duration that takes you past the end of the rollback, and verify that it _has_ finished.

For the positions, yet, you can use tester.getTopLeft and so forth to find coordinates.

**Comment by theankurkedia** [comment]
First, I have changed the name of `rollbackDuration` with `movementDuration` because I observed that while using the widget, this duration was also applicable when the fling is given and widget is dismissed.

**Comment by theankurkedia** [comment]
Second, as you suggested, I have modified tests. I have added 1 case here, in which the widget rolls back to original position (i.e. `movementDuration`). When we give I small nudge to it, the widget comes back to original place, but never goes out of `finder`, so the widget is always present. What I added was that while we give a fling to widget, it goes out of window after `movementDuration` and starts resizing. I have added checks whether we have a widget before completion of duration and after that duration and it is working fine. Although, I am still not that confident about this because the tests are still passing even when I change the default movementDuration.

**Comment by Hixie** [comment]
Thanks for your contribution!

If you like, you can submit additional tests in the future, to help ensure we don't accidentally regress part of your patch. (Anything that's not tested will eventually break, that's just a rule of software development.)

**Pull request merged**